Entrance Rando

Archez opened the pull request:
This PR is a continuation of #1052 rebased off the develop branch

I changed the entrance override check system to be a lookup table similar to the 3ds rando implementation system that operates on the `globalCtx->nextEntranceIndex` before the next scene loads. I added `randomizer_entrance.c` and `randomizer_grotto.c` which were pulled from the 3ds code and repurposed for SoH.

Most of the entrance override logic doesn't explicitly check for any of the entrance shuffle settings being on since the entrance table is initiated as unshuffled, so vanilla entrances will to go to their respective places. 

The 3ds code leveraged modifying the `gEntranceTable` to help make dealing with some edge cases easier, however on their end they don't have to worry about vanilla saves and rando saves working together. I opted to copy the original entrance table before modifications and restore when loading a vanilla save. In testing everything seemed fine when switching saves (vanilla save -> rando save -> vanilla save -> rando save), but I understand if this is a frowned upon approach.

Saves and spoiler logs are storing the original indexes, override indexes and blue warps for both directions (which will aid in supporting decoupled entrances later on). I couldn't think of a better way to save the information in the spoiler log that balances readability and accuracy.

I left in some code that is unused from these files in anticipation of continued work expected upstream from https://github.com/gamestabled/OoT3D_Randomizer/pull/566. I can strip that out for now and bring back what is needed when the upstream changes are done if that is preferred.

- [x] Support all dungeon shuffle options
- [x] Support overworld shuffle options
- [x] Support all interior shuffle options
- [x] Support grottos/graves shuffle options
- [x] Save/Read shuffle settings and index overrides from save file and spoiler log
- [x] Have dungeons accessible to child/adult always
- [x] Support grotto/grave randomized exits when using the debug menu to spawn inside a grotto/grave (this may also impacted mixed pools later on)
- [x] MQ dungeon entrance support

I'd like to continue adding the new entrance rando options from upstream, but I think this PR is in a really good playtest state so wanted to make it public now.

This is my first large contribution to a large C/C++ code base so if I didn't follow proper conventions, please let me know. Headers and externs are confusing!

---

### Note worthy changes:
* Deku Tree mouth is open for adult after showing Mido the sword&shield, or forest is set to open for Dungeon shuffle
* Fire Temple rocks are removed for child for Dungeon shuffle
* BotW rock blocking adult is removed if Song of Storms is played in the windmill for Dungeon shuffle
* GtG gate once opened as adult, will remain open permanently and for child too for Dungeon shuffle
* Impa's escort out of the castle garden will take link to the Castle grounds entrance near Market for Overworld shuffle
* Gerudo guards will always throw child link out to Gerudo valley even if he has hookshot (not restricted to ER)
* Gerudo guards will throw link out to the Fortress instead of the Valley if link is caught inside the Fortress (for adult and child) for Overworld shuffle
* There is an extra gate operator on the wasteland side for adult to open the gate on the other side for Overworld shuffle
* The shooting gallery man is removed when entering as the wrong age (not restricted to ER)
* Dying&continuing, using Sun's Song, or voiding out after returning from a grotto will respawn link next to that grotto, instead of at a default entrance (e.g. DMT has a special void out condition)
* A weather checking system to prevent the wrong weather/skybox conditions from being applied to a map
* A epona checking system to prevent link from spawing on epona automatically if he went through an entrance that doesn't support epona

![image](https://user-images.githubusercontent.com/13861068/197124919-dbc4df43-34cf-4972-9fad-eb993877c62a.png)

Co-authored-by: Ryanoberlein 102834664+Ryanoberlein@users.noreply.github.com

Ryanoberlein committed:
entrance rando seed gen

(cherry picked from commit 5c3aa0d3f297173a3a41f062c885c99d8e7f0f67)

Ryanoberlein committed:
fixed a build bug

(cherry picked from commit 69965b6b879694097c74d6065b17e2de0ff129e8)

Ryanoberlein committed:
entrance category added

(cherry picked from commit c73a6035d5b13811f90fb4b50db7eb2c0986d633)

Ryanoberlein committed:
fixed vcxproj files

(cherry picked from commit ca111fddd48da06ddda097bb4f35400c22331c1e)

Ryanoberlein committed:
fixed py file issue

(cherry picked from commit 9d78bb1b2026b19fb8dff3e851d3328886c2a9ee)

Ryanoberlein committed:
z_play integration

Currently broken, crashes on start.

(cherry picked from commit 07d2e43b04743dd97309716ba5b0842dc07c4175)

Ryanoberlein committed:
small logic fix

(cherry picked from commit 9a2ba83958b8ed6971a0e018ee91c9f4efeca172)

Ryanoberlein committed:
added more stuff to parse json

(cherry picked from commit 35932dfa5d951e8dcc45e17768a41f19e4dec54a)

Commit:
still not fully working but it seems at least some of the entrance replacements are being loaded

(cherry picked from commit 00b195f55b895cf6565b76fcdeff7c3dd62a4e4a)

Commit:
Merge branch 'ryan-test' into entrance-rando-small-fix, add note

(cherry picked from commit a88b0faa34499a05ddfdbf2fba056aa07d4b76a6)

Ryanoberlein committed:
working entrance rando

still needs human readable spoiler log and some imgui setting fixes

(cherry picked from commit e326c804e64281ca9b1c938e1f9dd92dab7374ee)

garrettjoecox committed:
Entirely replace gSaveContext.entranceIndex, and near the top of Gameplay_Init

(cherry picked from commit e60013db3cc904df298857f3d72949c86b0012ee)

Ryanoberlein committed:
Merge pull request #3 from garrettjoecox/entrance-rando-fixes

Entirely replace gSaveContext.entranceIndex, and near the top of Gameplay_Init

(cherry picked from commit 819f6695c1fad27fb3864c2b0794b5df671813c4)

Ryanoberlein committed:
human readable spoiler

(cherry picked from commit d2240eb9ab0afccb11b8ddf3afbc67813eb748be)

Ryanoberlein committed:
fixed grotto breaking bug

(cherry picked from commit 0890fc7d54f537976e3bffb76815da6f95408443)

Ryanoberlein committed:
actually fixed grotto bug (i hope)

(cherry picked from commit cdedccfbe652bf82281a1d1b1c459c2f342c51cb)

Ryanoberlein committed:
actually actually fixed grotto issue

(cherry picked from commit 59618a7333e359475a73599a3c2894476d52e9c2)

Ryanoberlein committed:
actually actually actually fixed grottos

(cherry picked from commit 0c15ddde7cf72b429f3d9cfef0fed13917cae036)

Ryanoberlein committed:
fairy fountain fix

(cherry picked from commit a95acaf6ddd88cbf91cb718862b2ff4c95eac69b)

Ryanoberlein committed:
fixed voidout/death/save&quit issue

(cherry picked from commit e034f028b3c1df26b6cfd1e94c261e13c43c6b38)

Ryanoberlein committed:
fixed lon lon

(cherry picked from commit 831285521ad1a814900a493e6b5a90296a4e0bdb)

Ryanoberlein committed:
fixed debug warping

(cherry picked from commit 33dfd12051f6a7e5a22d74795c17cdeffa02f793)

Ryanoberlein committed:
forgot a change

(cherry picked from commit 4c1459ee93e514addc52f11d0abd3c8933283bfa)

Ryanoberlein committed:
fixed sun song

(cherry picked from commit 19aedbca0803c0229b3708e25de0b9725b1d2435)

Ryanoberlein committed:
fixed minor github comment suggested changes

(cherry picked from commit 435d21ded573d8622b4d3facb568d16ff3eab311)

Ryanoberlein committed:
made firetemple and deku accesible for all ages

(cherry picked from commit 4da12e2974bc2e7b4edf0474aa3cec484c9636ca)

Ryanoberlein committed:
fixed farrores wind

(cherry picked from commit 60ec923b8d475e5b8765e31ca40d0e42c508846d)

Ryanoberlein committed:
fixed botw access

(cherry picked from commit ab6785180a145ad7ac89418695597d244566649a)

Ryanoberlein committed:
3drando parity botw logic

(cherry picked from commit c97778be31da790ba6d57340c7c43157bf332c67)

Ryanoberlein committed:
fixed hyrule market bridge visual bug

(cherry picked from commit e558efedd3aa9f1372e4e7dff8c9b781be710505)

Ryanoberlein committed:
mostly fixed boss warp

(cherry picked from commit 7ae998b2d53174ac469a04c98f33adc68497b224)

Ryanoberlein committed:
starting to work on horse cutscene issue

(cherry picked from commit 7d245d8661ec499986c7ae133353ef73d0ab275a)

aMannus committed:
Entrance rando UI cleanup

(cherry picked from commit e139ca275e993d6642868f15405d15a38e8f677d)

Ryanoberlein committed:
small fix

(cherry picked from commit e71f4222e2dac1d87d45017a591776d48dffada4)

Ryanoberlein committed:
made gerudo fortress accessible from desert entrance

(cherry picked from commit da7462d6e6788482caa49e4872b4b13f0bc916d5)

Ryanoberlein committed:
fixed epona cutscene issues

(cherry picked from commit 9a32b83fa16abbcc73e672e21a4144abee532db7)

Ryanoberlein committed:
fixed zora river decoupling

(cherry picked from commit 351edf3a3c791194820f99634f5e6584d678b4ac)

Archez committed:
couple formatting changes

Archez committed:
fix entrance rando gui setup

Archez committed:
improve spoiler log print for entrances

Archez committed:
camelcase entrance tracking flags

Archez committed:
move repeated entrance code into helper function

Archez committed:
Merge remote-tracking branch 'origin/develop' into continueEntranceRando

Archez committed:
fix bad func name

Archez committed:
use real entrances for better debug warp menu

Archez committed:
have console/cc comands respect entrance rando

Archez committed:
Merge remote-tracking branch 'origin/develop' into continueEntranceRando

Archez committed:
add missing child hc great fairy to better debug warp select

Archez committed:
add entrance/rando game logic files from 3ds source; modify for SoH

Archez committed:
change spoiler log and save context to store entrances better

Archez committed:
handle gorttos, voids, deaths, suns song and epona for entrance rando; remove old flags

Archez committed:
Merge remote-tracking branch 'origin/develop' into continueEntranceRando

Archez committed:
add some comments for entrance rando

Archez committed:
fix hyrule weather; try to preserve vanilla entrance table when starting vanilla saves after rando

Archez committed:
add support for shuffled ganons castle and all interiors

Archez committed:
remove noop call from file choose

Archez committed:
gerudo guards improvements on entrance rando; have gtg stay open for dungeon shuffled

Archez committed:
remove entrance tracker info for now

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_entrance.h:
```diff
@@ -0,0 +1,88 @@
+
+#include <stdint.h>
+
+//Entrance Table Data: https://wiki.cloudmodding.com/oot/Entrance_Table_(Data)
```
would you mind adding dates for when the data on the page was last updated and when it was accessed?  it seems that page is all pre-leak info so it should be all good, but it'll be good to have those dates documented just to be safe

Archez replied:
This link and most of the code in this file was pulled from 3ds rando. I didn't even think about the link potentially being bad. The commit that added this url on the 3ds side was June 22, 2021. So I can put that if you want

briaguya0 replied:
as long as we have the modified date of the page at that url in the comment we're golden
>  This page was last edited on 14 March 2020, at 21:47.

Archez committed:
fix jenkins build with missing externs; add comment for url published content data

Archez committed:
add missing return statement for dynamic exits

Archez committed:
Merge remote-tracking branch 'origin/develop' into continueEntranceRando

Archez committed:
handle shooting gallery man when the wrong age

Archez committed:
add header reference

Archez committed:
include header?

Archez committed:
commit missing func

Archez committed:
handle jabu fish drop cutscene override

Archez committed:
remove define for existing array count method

Archez committed:
Merge remote-tracking branch 'origin/develop' into continueEntranceRando

Archez committed:
fix better debug warp to load correct grottos

Archez committed:
fix randomized grotto/grave exits when using debug warp

briaguya0 commented on soh/src/overlays/actors/ovl_En_Syateki_Man/z_en_syateki_man.c:
```diff
@@ -154,6 +154,15 @@ void EnSyatekiMan_Init(Actor* thisx, GlobalContext* globalCtx) {
     s32 pad;
     EnSyatekiMan* this = (EnSyatekiMan*)thisx;
 
+    if (gSaveContext.n64ddFlag) {
```
do we want to do this always in rando? or just when doing entrance rando?

Archez replied:
3ds rando doesn't do a check for this, but I can add one here (for interior shuffle) if we want.

garrettjoecox replied:
Are there any glitches that involve child link going to adult shooting gallery or vice versa?

If so we probably should add an ER check here, if not this is probably fine.

Archez replied:
I'm not sure the usefulness, but I did find an example of child link clipping into adult shooting gallery. I added an interior shuffle check here now.

briaguya0 commented on soh/src/overlays/actors/ovl_En_Ishi/z_en_ishi.c:
```diff
@@ -330,6 +330,10 @@ void EnIshi_Init(Actor* thisx, GlobalContext* globalCtx) {
         return;
     }
+    // remove fire temple blocking rocks for child
+    if (gSaveContext.n64ddFlag && globalCtx->sceneNum == 0x061 && this->actor.params == (-255)) {
```
do we want to do this always in rando? or just when doing entrance rando?

Archez replied:
This, deku tree, botw, and gtg gate are all checking for dungeon entrance setting on 3ds rando. These are hold overs from the original entrance work, I can add those checks for them here.

briaguya0 commented on soh/src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:
```diff
@@ -93,6 +94,10 @@ void EnGe1_Init(Actor* thisx, GlobalContext* globalCtx) {
     s32 pad;
     EnGe1* this = (EnGe1*)thisx;
 
+    if (gSaveContext.n64ddFlag) {
```
do we want to do this always in rando? or just when doing entrance rando?

briaguya0 commented on soh/src/overlays/actors/ovl_Door_Ana/z_door_ana.c:
```diff
@@ -139,6 +139,11 @@ void DoorAna_WaitOpen(DoorAna* this, GlobalContext* globalCtx) {
             }
             globalCtx->nextEntranceIndex = entrances[destinationIdx];
+
+            if (gSaveContext.n64ddFlag) {
```
do we want to do this always in rando? or just when doing entrance rando?

Archez replied:
This and some of the functions in z_player/z_play that determine the entrance override, I decided to not explicitly check for entrance rando settings.

The entrance override logic uses vanilla indices for anything not shuffled and allows us to use a modified scene table. We could put entrance skips (like skip child stealth/ganon tower) into the entrance logic to have one place for that kind of stuff.

But if we don't want to do that, then I understand.

garrettjoecox replied:
> We could put entrance skips (like skip child stealth/ganon tower) into the entrance logic to have one place for that kind of stuff.

We want this 👍

briaguya0 commented on soh/src/overlays/actors/ovl_Bg_Treemouth/z_bg_treemouth.c:
```diff
@@ -73,7 +73,7 @@ void BgTreemouth_Init(Actor* thisx, GlobalContext* globalCtx) {
     if ((gSaveContext.sceneSetupIndex < 4) && !LINK_IS_ADULT) {
         BgTreemouth_SetupAction(this, func_808BC8B8);
-    } else if (LINK_IS_ADULT || (gSaveContext.sceneSetupIndex == 7)) {
+    } else if (LINK_IS_ADULT && !gSaveContext.n64ddFlag || (gSaveContext.sceneSetupIndex == 7)) {
```
for clarity/readability

```suggestion
    } else if ((LINK_IS_ADULT && !gSaveContext.n64ddFlag) || (gSaveContext.sceneSetupIndex == 7)) {
```

briaguya0 replied:
come do think of it, do we want deku as adult to be a thing in rando without entrance rando? not sure what logic thinks about it

Archez replied:
Same as fire temple rocks, will add the dungeon check here.

briaguya0 commented on soh/src/overlays/actors/ovl_Bg_Spot01_Idosoko/z_bg_spot01_idosoko.c:
```diff
@@ -47,7 +47,8 @@ void BgSpot01Idosoko_Init(Actor* thisx, GlobalContext* globalCtx) {
     CollisionHeader_GetVirtual(&gKakarikoBOTWStoneCol, &colHeader);
     this->dyna.bgId = DynaPoly_SetBgActor(globalCtx, &globalCtx->colCtx.dyna, &this->dyna.actor, colHeader);
-    if (!LINK_IS_ADULT) {
+    // remove bottom of the well rock for adult in rando if song of storms was played in windmill
```
do we want to do this always in rando? or just when doing entrance rando?

Archez replied:
Same as fire temple rocks, will add the dungeon check here.

briaguya0 reviewed (commented):
Overall this is looking beautiful! Super excited to see how it's coming along!

I left a few comments with questions, but assuming playtesting is going well this is super close to ready!

Archez committed:
add more comments and entrance setting checks

Archez committed:
Merge remote-tracking branch 'origin/develop' into continueEntranceRando

briaguya0 commented:
this should cover https://github.com/HarbourMasters/Shipwright/issues/717

Archez committed:
Merge remote-tracking branch 'origin/develop' into continueEntranceRando

Archez committed:
fix more hyrule weather; fix null pointer access; fix deku tree mouth

Archez committed:
fix 3ds logic enum for gc grotto entrance

Archez committed:
Merge remote-tracking branch 'origin/develop' into continueEntranceRando

Archez committed:
save entrance locations with sphere info in spoiler log

Archez committed:
simplify entrance rando grotto exit check for debug warp menu

Archez committed:
implement a better weather system checker and handle all weather combinations

Archez committed:
Merge remote-tracking branch 'origin/develop' into continueEntranceRando

Archez committed:
move entrance on/off RSK parsing in switch

Archez committed:
rename RSK values for entrance rando; add RSK values for rest of entrance rando work

Archez commented on soh/soh/Enhancements/randomizer/randomizerTypes.h:
```diff
@@ -1041,6 +1041,20 @@ typedef enum {
+    RSK_SHUFFLE_OVERWORLD_ENTRANCES,
+    RSK_SHUFFLE_INTERIOR_ENTRANCES,
+    RSK_SHUFFLE_GROTTO_ENTRANCES,
+    RSK_SHUFFLE_OWL_DROPS,
+    RSK_SHUFFLE_WARP_SONGS,
+    RSK_SHUFFLE_OVERWORLD_SPAWNS,
+    RSK_MIXED_ENTRANCE_POOLS,
+    RSK_MIX_DUNGEON_ENTRANCES,
+    RSK_MIX_OVERWORLD_ENTRANCES,
+    RSK_MIX_INTERIOR_ENTRANCES,
+    RSK_MIX_GROTTO_ENTRANCES,
+    RSK_DECOUPLED_ENTRANCES,
```
I threw in the rest of the entrance rando options in this enum to reserve their positions for when the v2 PR comes in later. If we don't like that, I can remove them.

Archez committed:
clean up entrance code; remove unused entrance tracker types

Archez committed:
include entrance header everywhere its used

Archez committed:
only open deku tree for adult if mido has seen sword and shield

Archez committed:
improve gerudo guard handling; limit gate opening to adult to match logic

Archez committed:
have impa bring link to castle grounds in ER; add quick comment

Archez committed:
Merge remote-tracking branch 'origin/develop' into continueEntranceRando

Archez committed:
simplify bluewarp overrides into function

garrettjoecox commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_death_mountain.cpp:
```diff
@@ -122,7 +122,7 @@ void AreaTable_Init_DeathMountain() {
                                       /*Glitched*/[]{return IsChild && Sticks && CanDoGlitch(GlitchType::QPA, GlitchDifficulty::ADVANCED);}}),
                   Entrance(GC_DARUNIAS_CHAMBER,  {[]{return (IsAdult && StopGCRollingGoronAsAdult) || GCDaruniasDoorOpenChild;}}),
-                  Entrance(GC_GROTTO,            {[]{return IsAdult && ((CanPlay(SongOfTime) && ((EffectiveHealth > 2) || CanUse(GORON_TUNIC) || CanUse(LONGSHOT) || CanUse(NAYRUS_LOVE))) || (EffectiveHealth > 1 && CanUse(GORON_TUNIC) && CanUse(HOOKSHOT)) || (CanUse(NAYRUS_LOVE) && CanUse(HOOKSHOT)));},
+                  Entrance(GC_GROTTO_PLATFORM,   {[]{return IsAdult && ((CanPlay(SongOfTime) && ((EffectiveHealth > 2) || CanUse(GORON_TUNIC) || CanUse(LONGSHOT) || CanUse(NAYRUS_LOVE))) || (EffectiveHealth > 1 && CanUse(GORON_TUNIC) && CanUse(HOOKSHOT)) || (CanUse(NAYRUS_LOVE) && CanUse(HOOKSHOT)));},
```
Why did this one need to change, and no others?

Archez replied:
This appears to have been a bug fix on the 3ds logic side that happened after our latest pull of 3ds logic (3+ months ago). So this seems to be the only thing that changed for this (was causing a grotto to not be included in ER for grotto shuffle).

garrettjoecox commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -2866,10 +2953,54 @@ void DrawRandoEditor(bool& open) {
+                    "- Windmill\n"
+                    "- Link's House\n"
+                    "- Temple of Time\n"
+                    "- Kakariko Potion Shop.\n"
```
```suggestion
                    "- Kakariko Potion Shop."
```

garrettjoecox commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -487,6 +492,12 @@ void Randomizer::LoadRequiredTrials(const char* spoilerFileName) {
+void Randomizer::LoadEntranceOverrides(const char* spoilerFileName, bool silent){
+    if (strcmp(spoilerFileName, "") != 0) {
+        ParseEntranceDataFile(spoilerFileName, silent);
+    }
```
Do we not need to augment gSaveContext here?

Archez replied:
The Parse here sets up gSaveContext on save_init, and then SaveManager ensures its saved/loaded correctly in all other scenarios.

This was the main thing I was struggling trying to understand what our preferred patterns are (hence my confusion with the recent discussion for the MQ dungeon count bug). Some functions maintain an array/map in the randomizer instance with SaveManager, while some things just rely solely on gSaveContext.

garrettjoecox commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -1092,6 +1124,52 @@ void Randomizer::ParseItemLocationsFile(const char* spoilerFileName, bool silent
+            for (auto entranceIt = entranceJson.begin(); entranceIt != entranceJson.end(); ++entranceIt) {
+                if (entranceIt.key() == "index") {
+                    gSaveContext.entranceOverrides[i].index = entranceIt.value();
+                } else if (entranceIt.key() == "destination") {
```
Any reason this is an `else if` and the rest here are separate `if` conditions?

Archez replied:
ah must have been copy pasta

Archez replied:
Fixed

garrettjoecox commented on soh/soh/Enhancements/randomizer/randomizer_grotto.c:
```diff
@@ -0,0 +1,256 @@
+    {.entranceIndex = 0x0129, .room = 0x00, .angle = 0x4000, .pos = {.x =   380.521f,  .y =   333.0f, .z = -1560.74f}},  // GF Storms Grotto -> Gerudo Fortress
+    {.entranceIndex = 0x022D, .room = 0x00, .angle = 0x9555, .pos = {.x = -1326.34f,   .y =    15.0f, .z =  -983.994f}}, // GV Storms Grotto -> GV Fortress Side
+    {.entranceIndex = 0x0117, .room = 0x00, .angle = 0x8000, .pos = {.x =   291.513f,  .y =  -555.0f, .z =  1478.39f}},  // GV Octorok Grotto -> GV Grotto Ledge
+    {.entranceIndex = 0x01A9, .room = 0x06, .angle = 0x4000, .pos = {.x =   109.281f,  .y =   -20.0f, .z = -1601.42f}},  // Deku Theater -> LW Beyond Mido
```
Side note: Part of me wonders if we could use this and get rid of this data in the better z select screen.

garrettjoecox commented on soh/soh/SaveManager.cpp:
```diff
@@ -139,6 +139,16 @@ void SaveManager::LoadRandomizerVersion2() {
         });
     });
 
+    SaveManager::Instance->LoadArray("entrances", ARRAY_COUNT(gSaveContext.entranceOverrides), [&](size_t i) {
```
Will `gSaveContext.entranceOverrides` contain _all_ entrances, or only overriden entrances? 

EG if I didn't randomize entrances I would expect this to be empty, or if I only shuffled grottos I would expect to only see grottos here.

Archez replied:
This will only contain shuffled entrances as it is coded now (both directions, so when decoupled entrances == ON would consume the same amount of indexes in this array as == OFF)

garrettjoecox commented on soh/src/overlays/actors/ovl_Bg_Spot01_Idosoko/z_bg_spot01_idosoko.c:
```diff
@@ -47,7 +47,10 @@ void BgSpot01Idosoko_Init(Actor* thisx, GlobalContext* globalCtx) {
+    // If dungeon entrance randomizer is on, remove the well stone as adult Link when
+    // child Link has drained the water to the well
+    if (!LINK_IS_ADULT || gSaveContext.n64ddFlag && Randomizer_GetSettingValue(RSK_SHUFFLE_DUNGEON_ENTRANCES) &&
+        gSaveContext.eventChkInf[6] &0x80) {
```
Use Flags_GetEventChkInf and related methods here and wherever possible

garrettjoecox commented on soh/soh/Enhancements/randomizer/randomizer_entrance.c:
```diff
@@ -0,0 +1,497 @@
+
+// Certain weather states don't clear normally in entrance rando, so we need to reset and re-apply
+// the correct weather state based on the current entrance and scene
+void Entrance_CheckWeatherState() {
```
This, and probably all other new methods with "Check" in their name should be renamed. I think they were inherited from 3drando like this but their current name does not tell me that they would actually modify anything, rather I would think would just validate and throw an error or return a bool.

garrettjoecox commented on soh/src/overlays/actors/ovl_Bg_Spot12_Saku/z_bg_spot12_saku.c:
```diff
@@ -125,6 +131,12 @@ void func_808B37AC(BgSpot12Saku* this, GlobalContext* globalCtx) {
+    // If ER is on, when the guard opens the GtG gate its permanent flag will be set.
+    if (gSaveContext.n64ddFlag && Randomizer_GetSettingValue(RSK_SHUFFLE_DUNGEON_ENTRANCES) &&
+        globalCtx->actorCtx.flags.tempSwch & 0x04000000) {
+        globalCtx->actorCtx.flags.swch |= 0x00000004;
```
Likewise here with flags `Flags_SetSwitch` if possible

garrettjoecox commented on soh/src/overlays/actors/ovl_Demo_Im/z_demo_im.c:
```diff
@@ -867,7 +867,12 @@ void func_80986B2C(GlobalContext* globalCtx) {
         Player* player = GET_PLAYER(globalCtx);
 
-        globalCtx->nextEntranceIndex = 0xCD;
+        // In entrance rando have impa bring link back to the front of castle grounds
```
thank you for the comments on all of these 👍

garrettjoecox commented on soh/src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:
```diff
@@ -93,6 +94,18 @@ void EnGe1_Init(Actor* thisx, GlobalContext* globalCtx) {
+    // When spawning the gate operator, also spawn an extra gate operator on the wasteland side
+    if (gSaveContext.n64ddFlag && Randomizer_GetSettingValue(RSK_SHUFFLE_GERUDO_MEMBERSHIP_CARD) &&
+        Randomizer_GetSettingValue(RSK_SHUFFLE_OVERWORLD_ENTRANCES) && (this->actor.params & 0xFF) == GE1_TYPE_GATE_OPERATOR) {
+        Actor_Spawn(&globalCtx->actorCtx, globalCtx, ACTOR_EN_GE1, -1358.0f, 88.0f, -3018.0f, 0, 0x95B0, 0, 0x0302);
```
Would be good to either explain or use an expression to show how you went from `0xFF` (used below) to `0x0302`

Archez replied:
This came from 3ds rando, I don't know where the `0x0300` came from but `0x0002` is from the custom guard type added, I can use an OR to at least make that clear.

garrettjoecox commented on soh/src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:
```diff
@@ -247,6 +265,10 @@ void EnGe1_KickPlayer(EnGe1* this, GlobalContext* globalCtx) {
             globalCtx->nextEntranceIndex = 0x3B4;
         }
 
+        if (gSaveContext.n64ddFlag) {
```
Similar to Bria's other comment, do we want these always in rando? Or just with entrance rando

Archez replied:
This one is special, it does an ER check inside for the ER logic, but also makes it so child link with hookshot does not get put into jail causing a soft lock (technically SoH supports child equip swap for hookshot), so I figured this was a worthy QoL for rando in general.

garrettjoecox commented on soh/src/overlays/actors/ovl_En_Ishi/z_en_ishi.c:
```diff
@@ -330,6 +330,12 @@ void EnIshi_Init(Actor* thisx, GlobalContext* globalCtx) {
+    // If dungeon entrance randomizer is on, remove the grey boulders that normally
+    // block child Link from reaching the Fire Temple entrance.
+    if (type == ROCK_LARGE && gSaveContext.n64ddFlag && Randomizer_GetSettingValue(RSK_SHUFFLE_DUNGEON_ENTRANCES) &&
+        globalCtx->sceneNum == 0x061) { // Death Mountain Creater
```
Does this not remove all rocks in DMC? EG, the ones blocking fairy fountain and the DMC scrub grotto

Archez replied:
ROCK_LARGE is silver gauntlet rocks, so it doesn't affect the grotto or fairy fountain that are hammer rocks (this is how 3ds rando does it).

garrettjoecox commented on soh/src/overlays/gamestates/ovl_file_choose/z_file_choose.c:
```diff
@@ -298,6 +298,7 @@ void FileChoose_UpdateMainMenu(GameState* thisx) {
         Randomizer_LoadItemLocations(fileLoc, silent);
         Randomizer_LoadMerchantMessages(fileLoc);
         Randomizer_LoadMasterQuestDungeons(fileLoc);
+        Randomizer_LoadEntranceOverrides(fileLoc, silent);
```
Do we not need a `Randomizer_LoadEntranceOverrides("");` near the top of `FileChoose_LoadGame`?

Archez replied:
Since, as its coded now, the entrance overrides are pulled from gSaveContext during sram_opensave, the load function is a noop here so I didn't include it. But I probably should add it here for posterity sake and to work with your fast file select changes (ER works fine as is with FFS since gSaveContext isn't augmented)

Archez replied:
Ok I added the load call into the `FileChoose_LoadGame` function. Currently its a noop, but at least if it ever changes, we don't have to worry about it being missing here.

garrettjoecox commented on soh/src/overlays/gamestates/ovl_select/z_select.c:
```diff
@@ -623,34 +637,34 @@ static BetterSceneSelectEntry sBetterScenes[] = {
+    { 0x05A8, 0x018D, 0, 0xE4, 0x51, { -7873.0,  -300.0,  6916.0 }}, // Hyrule Field -> HF Cow Grotto
+    { 0x05FC, 0x01B9, 0, 0xF8, 0x60, {  -678.0,  1946.0,  -284.0 }}, // Death Mountain Summit -> DMT Cow Grotto
+    { 0x05AC, 0x0117, 0, 0xF2, 0x5A, {   271.0,  -555.0,  1465.0 }}, // GV Grotto Ledge -> GV Octorok Grotto
+    { 0x05C0, 0x00CD, 0, 0xE1, 0x51, { -4945.0,  -300.0,  2841.0 }}, // Hyrule Field -> HF Tektite Grotto
```
🔥

garrettjoecox reviewed (commented):
Terrific work! Left some comments throughout and have two more high level comments;

1. I noticed a few more places where we are only checking `gSaveContext.n64ddFlag` and not ER for things. If the reasoning for all of these is the reason you explained to bria where we might eventually move our hand crafted entrance skips to using these systems then I'm totally fine with those, otherwise we should add an ER check to them. Would you do a once over on each of those cases to make sure?

2. It looks like the meat of the new files (randomizer_entrance/randomizer_grotto) were pulled from 3drando, I know it's probably been a while since you messed with these but it would be good to add a comment at the top of them linking to their origin and maybe a very brief summary of high level changes that were made to adapt them to SoH

Archez committed:
add feedback suggestions and slight cleanup

Archez committed:
add comments about borrowing 3ds code; rename _check funcs to be more clear

Archez committed:
Merge remote-tracking branch 'origin/develop' into continueEntranceRando

Archez committed:
remove ganons tower skip from entrance logic since its handled elsewhere right now

Archez commented:
> 1. I noticed a few more places where we are only checking gSaveContext.n64ddFlag and not ER for things. If the reasoning for all of these is the reason you explained to bria where we might eventually move our hand crafted entrance skips to using these systems then I'm totally fine with those, otherwise we should add an ER check to them. Would you do a once over on each of those cases to make sure?
> 2. It looks like the meat of the new files (randomizer_entrance/randomizer_grotto) were pulled from 3drando, I know it's probably been a while since you messed with these but it would be good to add a comment at the top of them linking to their origin and maybe a very brief summary of high level changes that were made to adapt them to SoH

I addressed most of the inline comments.

For 1. I added ER checks to the ones that make sense needing it (mostly anything grotto related since a normal entrance skip would never need to deal with the custom grotto logic from ER).
For 2. I added some comments at the beginning of those files.

~~The linux build seems to be failing due to a git clone that can't complete.~~

Archez committed:
fix exiting hyrule castle courtyard as child link; move link spawn when exiting ganons castle randomized dungeon

Archez committed:
clarify shuffle dungeon description

Archez committed:
Merge remote-tracking branch 'origin/develop' into continueEntranceRando

garrettjoecox reviewed (approved):
Alrighty you've got my approval here. Great work!

briaguya0 commented on soh/soh/Enhancements/randomizer/3drando/spoiler_log.cpp:
```diff
@@ -523,23 +543,16 @@ static void WritePlaythrough() {
+  for (uint32_t i = 0; i < playthroughEntrances.size(); ++i) {
+    auto sphereNum = std::to_string(i);
+    std::string sphereString = "sphere ";
+    if (sphereNum.length() == 1) sphereString += "0";
```
nitpick: i think just doing `i < 10` might be nicer here

Archez replied:
updated

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -2488,22 +2564,33 @@ void GenerateRandomizerImgui() {
-        while (getline(excludedLocationStringStream, excludedLocationString, ',')) {
-            excludedLocations.insert((RandomizerCheck)std::stoi(excludedLocationString));
-        }
+    // Still used in 3DS logic, but not in our UI. Instead, enable if any of the entrance rando options are enabled.
```
nitpick, `RSK_SHUFFLE_ENTRANCES` is used in SoH logic too

Archez replied:
clarified comment

briaguya0 commented on soh/src/code/z_sram.c:
```diff
@@ -196,6 +197,19 @@ void Sram_OpenSave() {
     }
 
+    // Setup the entrance override table for rando
+    if (gSaveContext.n64ddFlag) {
```
do we need to init the override table if entrance shuffle is off?

Archez replied:
Here, and the two below are the main uses of the "entrance override" logic, and relates to the early conversations above with @garrettjoecox. It achieves two things:
1. Allows us to modify the games entrance table for entrance skips or modifications (like skip zelda stealth)
2. Setups up the entrance shuffle table (uses vanilla data if un-shuffled)

The thought is point 1 could be useful outside of entrance shuffle, and since the shuffle table is initialized with vanilla data, it should be safe to use.

Lets decide what we want to do with this PR. We could add the entrance shuffle check here for now, and re-visit it later when we figure out how we want to take advantage of a modified entrance table.

briaguya0 replied:
as long as initializing the table doesn't break anything in non-entrance rando i say let's keep it, it'll be nice to use it to clean up some of the other entrance override stuff we're doing in hacky ways in the future

briaguya0 commented on soh/src/overlays/actors/ovl_Bg_Treemouth/z_bg_treemouth.c:
```diff
@@ -73,7 +73,9 @@ void BgTreemouth_Init(Actor* thisx, GlobalContext* globalCtx) {
     if ((gSaveContext.sceneSetupIndex < 4) && !LINK_IS_ADULT) {
         BgTreemouth_SetupAction(this, func_808BC8B8);
-    } else if (LINK_IS_ADULT || (gSaveContext.sceneSetupIndex == 7)) {
+    // If dungeon entrance randomizer is on, keep the tree mouth open link is adult and sword and shield are shown to mido
```
this comment feels weird, if nothing else there should be a `if` or `when` in there

Archez replied:
clarified comment

briaguya0 commented on soh/src/overlays/actors/ovl_Demo_Im/z_demo_im.c:
```diff
@@ -867,7 +867,12 @@ void func_80986B2C(GlobalContext* globalCtx) {
-        globalCtx->nextEntranceIndex = 0xCD;
+        // In entrance rando have impa bring link back to the front of castle grounds
+        if (gSaveContext.n64ddFlag && Randomizer_GetSettingValue(RSK_SHUFFLE_OVERWORLD_ENTRANCES)) {
+            globalCtx->nextEntranceIndex = 0x0138;
```
idea for a future PR: replace magic number entrance indices with enum values

briaguya0 commented on soh/src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:
```diff
@@ -93,6 +94,20 @@ void EnGe1_Init(Actor* thisx, GlobalContext* globalCtx) {
     s32 pad;
     EnGe1* this = (EnGe1*)thisx;
 
+    // When spawning the gate operator, also spawn an extra gate operator on the wasteland side
```
is this something we were missing from non-entrance rando?

Archez replied:
Yes, 3ds code places the guard there if the gerudo card is shuffled or entrance rando, and is accounted for in logic too

briaguya0 commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -4210,14 +4211,26 @@ s32 func_80839034(GlobalContext* globalCtx, Player* this, CollisionPoly* poly, u
                 globalCtx->nextEntranceIndex = globalCtx->setupExitList[sp3C - 1];
+
+                // Main override for entrance rando and entrance skips
+                if (gSaveContext.n64ddFlag) {
```
should we be checking for entrance rando here?

Archez replied:
Answered here: https://github.com/HarbourMasters/Shipwright/pull/1760#discussion_r1010976707

briaguya0 commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -4210,14 +4211,26 @@ s32 func_80839034(GlobalContext* globalCtx, Player* this, CollisionPoly* poly, u
-                    globalCtx->nextEntranceIndex =
-                        D_808544F8[D_80854514[globalCtx->nextEntranceIndex - 0x7FF9] + globalCtx->curSpawn];
+                    // handle dynamic exits
+                    if (gSaveContext.n64ddFlag) {
```
should we be checking for entrance rando here?

Archez replied:
Answered here: https://github.com/HarbourMasters/Shipwright/pull/1760#discussion_r1010976707

briaguya0 reviewed (commented):
this is super close to good to go

left couple little nitpicks and questions in there, but overall this is super exciting to see!

Archez committed:
ckarify comments from feedback

Archez committed:
Merge remote-tracking branch 'origin/develop' into continueEntranceRando

dcvz commented:
Was taking a look at this PR: @Archez seems like there's still a couple of open comments

Archez committed:
Merge remote-tracking branch 'origin/develop' into continueEntranceRando

Archez committed:
rename to use playstate/play

Archez committed:
only despawn the shooting gallery man for interior shuffle

dcvz reviewed (approved):
Great work on this @Archez! Thank you for addressing all the comments and adding more documentation in the code.
